martha/martha_v3.js

```diff
@@ -40,38 +160,75 @@ function marthaV3Handler(req, res) {
         res.status(BAD_REQUEST_ERROR_CODE).send(failureResponse);
         return;
     }
+    const parsedUrl = url.parse(dataObjectUri);
+    if (!parsedUrl.host || (!parsedUrl.path && parsedUrl.host.toLowerCase().startsWith('dg.'))) {
+        console.error(`"${url.format(parsedUrl)}" is missing a host and/or a path.`);
+        const failureResponse = new FailureResponse(BAD_REQUEST_ERROR_CODE, `"${dataObjectUri}" is not a properly-formatted URI.`);
+        res.status(BAD_REQUEST_ERROR_CODE).send(failureResponse);
+        return;
+    }
+    const {drsUrl, sendAuth, bondUrl, responseParser} = determineDrsType(parsedUrl);
+    console.log(`Converting DRS URI to HTTPS: ${dataObjectUri} -> ${drsUrl}`);
 
-    console.log(`Received URL '${dataObjectUri}' from IP '${req.ip}'`);
-    let dataObjectResolutionUrl;
+    let response;
     try {
-        dataObjectResolutionUrl = dataObjectUriToHttps(dataObjectUri);
+        if (sendAuth) {
+            response = await apiAdapter.getJsonFrom(drsUrl, auth);
+        } else {
+            response = await apiAdapter.getJsonFrom(drsUrl);
+        }
     } catch (err) {
+        console.log('Received error while resolving DRS URL.');
         console.error(err);
-        const failureResponse = new FailureResponse(BAD_REQUEST_ERROR_CODE, `The specified URL '${dataObjectUri}' is invalid`);
-        res.status(BAD_REQUEST_ERROR_CODE).send(failureResponse);
+
+        const errorStatusCode = err.status;
+        if (typeof errorStatusCode === 'undefined') {
+            const failureResponse = new FailureResponse(SERVER_ERROR_CODE, `Received error while resolving DRS URL. ${err.message}`);
+            res.status(SERVER_ERROR_CODE).send(failureResponse);
+        } else {
+            res.status(errorStatusCode).send(err);
+        }
         return;
     }
 
-    const bondProvider = determineBondProvider(dataObjectUri);
-    const dataObjectPromise = getDataObjectMetadata(dataObjectResolutionUrl, auth, bondProvider);
-    const bondPromise = maybeTalkToBond(req, bondProvider);
+    let drsResponse;
+    try {
+        drsResponse = responseParser(response);
+    } catch (err) {
+        console.log('Received error while parsing response from DRS URL.');
+        console.error(err);
+
+        const errorStatusCode = err.status;
+        if (typeof errorStatusCode === 'undefined') {
+            const failureResponse = new FailureResponse(SERVER_ERROR_CODE, `Received error while parsing response from DRS URL. ${err.message}`);
+            res.status(SERVER_ERROR_CODE).send(failureResponse);
+        } else {
+            res.status(errorStatusCode).send(err);
+        }
+        return;
```

Comment:
This looks an opportunity to write some DRY code. This block of code in the `catch` block is pretty heavily copy and pasted from elsewhere in this function.  This looks like a function that could be refactored out into its own function which takes an error object and a message to do the same work in multiple places.

Reply:
This was discussed in earlier comments: see https://github.com/broadinstitute/martha/pull/173#pullrequestreview-450931894 and 
https://github.com/broadinstitute/martha/pull/173#issuecomment-660825418 .

Reply:
I saw Saloni's suggestion and Khalid's response.  I would be ok with Saloni's suggestion, but am also fine with your response. What I was suggesting is something slightly different.  Create this helper function:

```function logError(err, message) {
        console.log(message);
        console.error(err);

        const errorStatusCode = err.status;
        if (typeof errorStatusCode === 'undefined') {
            const failureResponse = new FailureResponse(SERVER_ERROR_CODE, `{$message} ${err.message}`);
            res.status(SERVER_ERROR_CODE).send(failureResponse);
        } else {
            res.status(errorStatusCode).send(err);
        }
        return;
}```

then use it in the main v3 endpoint:

```
try {
   # first chunk of work
} catch (err) {
	logError(err, "message 1")
}

try {
   # 2nd chunk of work
} catch (err) {
	logError(err, "message 2")
}
...
```

Some food for thought to get the spirit of what Saloni was requesting while keeping the different error messages that both you and Khalid think are important to keep.